### ISSUE TITLE
Support get_c_str() in On Demand

### ISSUE THREAD

**Issue by jkeiser** [simdjson/simdjson#1214]
string_view doesn't actually promise that a string is null-terminated, but we do null-terminate strings in ondemand. We should support a `const char *` cast and `.get_c_str()` in ondemand::value and ondemand::document.

**Comment by Tachi107** [comment]
> string_view doesn't actually promise that a string is null-terminated, but we do null-terminate strings in ondemand

This doesn't seem the case, as `object["key"].get_string().value().data()` doesn't return a NULL-terminated string

**Comment by lemire** [comment]
I believe that's correct. We do not null-terminate the strings. At the very least, I see nowhere a promise that the string will be null terminated.

Importantly, a null character is allowed anywhere within a string in JSON, so null termination is problematic. In JSON, a null character is just like any other ASCII character.

(This is somewhat annoying. The first release of simdjson assumed that strings did not contain null characters and we got a lot of complains about it...)

**Comment by Tachi107** [comment]
Yes, NULL-terminating the returned string without allocating new memory could be problematic for simdjson, as as far as I know you just iterate over the json content without modifying it. 

If this is the case, I think that this thread could be closed. Creating null-terminated strings in user code is pretty straightforward if you don't mind the extra allocation.

**Comment by lemire** [comment]
@Tachi107  Though we give you raw access to the bytes from the original source, we actually have to process the input and copy it. So the string_view you get is pointing at a processed copy. We definitely could pad it with a null character if we wanted to. We probably did so at some point, hence John’s comment.

**Comment by jkeiser** [comment]
I'm a little surprised we don't null terminate anymore when we copy / decode the string; I didn't recall removing that code, though it's been some time now.

**Comment by lemire** [comment]
Presumably, it would be done in the following function:

https://github.com/simdjson/simdjson/blob/1aa1d537fe6a588b129f6c81940b32da4ae4d9fc/include/simdjson/generic/stringparsing.h#L76-L124


I do not see any indication that we null terminate or that we intended to null terminate. This can be changed, of course.

**Comment by jkeiser** [comment]
You're right! I think I probably did shave it off to see if anyone would complain.

I'm pretty sure the lack of get_c_str() is just an omission I never remembered to slam in there. It could easily be done by calling the non terminating version and then adding a null character afterwards.

I would add the method no matter what, so that people have a way to be explicit. However, if zero padding's cost is negligible (and it should be) I think we should do it unconditionally, even when you convert to string_view. We are moving in C++, after all, and I suspect a huge number of people assume zero padding everywhere (and are generally right because most apis append the null even when they don't strictly need to).
